packages/react-compose-next/src/compose.ts

```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { ComposedComponent, ComposeOptions, ComposePreparedOptions } from './types';
+
+const defaultComposeOptions: ComposePreparedOptions = {
+  className: process.env.NODE_ENV === 'production' ? '' : 'no-classname-🙉',
+  displayNames: [],
+
+  mapPropsToStylesPropsChain: [],
+
+  handledProps: [],
+  overrideStyles: false
+};
+
+function computeDisplayNames(
+  options: ComposeOptions<any, any, any>,
+  inputDisplayName?: string,
+  inputOptions?: ComposePreparedOptions<any, any, any, any>
+): string[] {
+  if (options.overrideStyles) {
+    return [options.displayName || inputDisplayName].filter(Boolean) as string[];
+  }
+
+  // To support styles composition we need to properly pick up display names
+  if (inputOptions) {
+    return options.displayName ? inputOptions.displayNames.concat(options.displayName) : inputOptions.displayNames;
+  }
+
+  return [inputDisplayName, options.displayName].filter(Boolean) as string[];
+}
+
+function compose<
+  InputProps extends Record<string, any>,
+  InputStylesProps extends Record<string, any>,
+  ParentProps extends Record<string, any>,
+  ParentStylesProps extends Record<string, any>
+>(
+  InputComponent: React.FunctionComponent<ParentProps>,
```

Comment:
This will not accept a `forwardRef`, I will test it separately and will make a follow up